dbms/src/Storages/StorageMySQL.cpp

```diff
@@ -0,0 +1,187 @@
+    String name = where.getID();
+    if ((name == "Function_and") || (name == "Function_or"))
+    {
+        const ASTFunction * and_expression = typeid_cast<const ASTFunction *>(&where);
+        for (const auto & expr: and_expression->arguments->children)
+        {
+            if (!isCompatible(*expr.get()))
+                return false;
+        }
+    }
+    else if (name == "Function_equals" || (name == "Function_notEquals") || (name == "Function_greater") || (name == "Function_less")
+        || (name == "Function_lessOrEquals") || (name == "Function_greaterOrEquals"))
+    {
+        const ASTFunction * function = typeid_cast<const ASTFunction *>(&where);
+        const auto & children = function->arguments->children;
+        if ((children.size() != 2) || !isCompatible(*children[0].get()) || !isCompatible(*children[1].get()))
+            return false;
+    }
+    else if (name == "Function_not")
+    {
+        const ASTFunction * function = typeid_cast<const ASTFunction *>(&where);
+        const auto & children = function->arguments->children;
+        if ((children.size() != 1) || !isCompatible(*children[0].get()))
+            return false;
+    }
+    else if ((strncmp(name.c_str(), "Identifier_", 11) == 0) || (strncmp(name.c_str(), "Literal_", 8) == 0))
+        return true;
+    else
+        return false;
+    return true;
+}
+
+void dumpWhere(const IAST & where, std::stringstream & stream)
+{
+    IAST::FormatSettings s(stream, false, true);
+    IAST::FormatState state;
+    IAST::FormatStateStacked frame;
+    where.formatImpl(s, state, frame);
+}
+
+/** Function puts to stream compatible expressions of where statement.
+  * Compatible expressions are logical expressions on logical expressions or comparisons between fields or constants
+  */
+
+void filterWhere(const IAST & where, std::stringstream & stream)
+{
+    String name = where.getID();
+    if (name == "Function_and")
+    {
+        const ASTFunction * and_expression = typeid_cast<const ASTFunction *>(&where);
+        bool first = true;
+        for (const auto & expr: and_expression->arguments->children)
+        {
+            if (isCompatible(*expr.get()))
+            {
+                if (!first)
+                    stream << " AND ";
+                first = false;
+                dumpWhere(*expr.get(), stream);
+            }
+        }
+    }
+    else
+    {
+        if (isCompatible(where))
+        {
+            dumpWhere(where, stream);
+        }
+    }
+}
+
+/** Function analyze query builds select query of all used columns in query_info from table set by table_name parameter with where expression from filtered query_info.
+  * Also it is builds sample_block with all columns, where types are found in column_map
+  */
+
+std::string analyzeQuery(const SelectQueryInfo & query_info, const Context & context, std::string table_name, NamesAndTypesListPtr columns, google::dense_hash_map<std::string, DataTypePtr> & column_map, Block & sample_block)
```

Comment:
`const std::string & table_name`
`const NamesAndTypesListPtr & columns`
`const google::dense_hash_map<std::string, DataTypePtr> & column_map`
Probably, it's better to split function in two parts: create sample_block and query separately (also to avoid output argument).